File: backend/app/schemas/maintenance_work.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from app.models.maintenance_work import WorkStatus

# Maintenance Step schema
class MaintenanceStep(BaseModel):
    step: int
    description: str
    completed: bool
    completedAt: Optional[datetime] = None

# Request schemas
class MaintenanceWorkCreate(BaseModel):
    requestId: int
    startedAt: Optional[datetime] = None
    workDescription: Optional[str] = None
    maintenanceSteps: Optional[List[MaintenanceStep]] = None
    
    class Config:
        extra = "forbid"

class MaintenanceWorkUpdate(BaseModel):
    startedAt: Optional[datetime] = None
    workDescription: Optional[str] = None
    maintenanceSteps: Optional[List[MaintenanceStep]] = None
    completedAt: Optional[datetime] = None
    status: Optional[WorkStatus] = None

# Progress tracking schemas
class MaintenanceWorkStart(BaseModel):
    """Schema for starting maintenance work"""
    workDescription: Optional[str] = None
    maintenanceSteps: Optional[List[MaintenanceStep]] = None
    
    class Config:
        extra = "forbid"

class MaintenanceWorkProgressUpdate(BaseModel):
    """Schema for updating maintenance work progress"""
    maintenanceSteps: List[MaintenanceStep] = Field(..., description="Updated maintenance steps")
    
    class Config:
        extra = "forbid"

class MaintenanceWorkComplete(BaseModel):
    """Schema for completing maintenance work"""
    workDescription: str = Field(..., min_length=1, description="Work description is required when completing")
    maintenanceSteps: Optional[List[MaintenanceStep]] = None
    notes: Optional[str] = Field(None, description="Optional completion notes")
    
    class Config:
        extra = "forbid"

class MaintenanceWorkResponse(BaseModel):
    id: int
    requestId: int
    assignedToId: int
    machineId: int
    workDescription: str
    status: WorkStatus
    startTime: Optional[datetime] = None
    endTime: Optional[datetime] = None
    estimatedHours: Optional[float] = None
    actualHours: Optional[float] = None
    laborCost: Optional[float] = None
    materialCost: Optional[float] = None
    totalCost: Optional[float] = None
    failureCodeId: Optional[int] = None
    maintenanceTypeId: Optional[int] = None
    maintenanceSteps: Optional[List[MaintenanceStep]] = None
    createdAt: datetime
    updatedAt: datetime
    
    class Config:
        from_attributes = True

File: frontend/src/lib/api/maintenance-work.ts
import { apiClient } from '../api-client';
import { WorkStatus } from '../types';

export interface MaintenanceStep {
  step: number;
  description: string;
  completed: boolean;
  completedAt?: string;
}

export interface MaintenanceWork {
  id: number;
  requestId: number;
  assignedToId: number;
  machineId: number;
  workDescription: string;
  status: WorkStatus;
  startTime?: string;
  endTime?: string;
  estimatedHours?: number;
  actualHours?: number;
  laborCost?: number;
  materialCost?: number;
  totalCost?: number;
  failureCodeId?: number;
  maintenanceTypeId?: number;
  maintenanceSteps?: MaintenanceStep[];
  createdAt: string;
  updatedAt: string;
}

export interface MaintenanceWorkCreate {
  requestId: number;
  startedAt?: string;
  workDescription?: string;
  maintenanceSteps?: MaintenanceStep[];
}

export interface MaintenanceWorkUpdate {
  startedAt?: string;
  workDescription?: string;
  maintenanceSteps?: MaintenanceStep[];
  completedAt?: string;
  status?: WorkStatus;
}

// Progress tracking schemas
export interface MaintenanceWorkStart {
  workDescription?: string;
  maintenanceSteps?: MaintenanceStep[];
}

export interface MaintenanceWorkProgressUpdate {
  maintenanceSteps: MaintenanceStep[];
}

export interface MaintenanceWorkComplete {
  workDescription: string;
  maintenanceSteps?: MaintenanceStep[];
  notes?: string;
}

export const maintenanceWorkApi = {
  // Get maintenance work by ID
  getWork: async (workId: number): Promise<MaintenanceWork> => {
    const response = await apiClient.get(`/maintenance-work/${workId}`);
    return response.data;
  },

  // Get maintenance work by request ID
  getWorkByRequest: async (requestId: number): Promise<MaintenanceWork | null> => {
    try {
      const response = await apiClient.get(`/maintenance-work/by-request/${requestId}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 204 || error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  },

  // Create maintenance work
  createWork: async (data: MaintenanceWorkCreate): Promise<MaintenanceWork> => {
    const response = await apiClient.post('/maintenance-work', data);
    return response.data;
  },

  // Update maintenance work (generic patch)
  updateWork: async (workId: number, data: MaintenanceWorkUpdate): Promise<MaintenanceWork> => {
    const response = await apiClient.patch(`/maintenance-work/${workId}`, data);
    return response.data;
  },

  // Start work
  startWork: async (workId: number, data: MaintenanceWorkStart): Promise<MaintenanceWork> => {
    const response = await apiClient.patch(`/maintenance-work/${workId}/start`, data);
    return response.data;
  },

  // Update progress (dedicated endpoint)
  updateProgress: async (workId: number, data: MaintenanceWorkProgressUpdate): Promise<MaintenanceWork> => {
    const response = await apiClient.patch(`/maintenance-work/${workId}/update-progress`, data);
    return response.data;
  },

  // Complete work
  completeWork: async (workId: number, data: MaintenanceWorkComplete): Promise<MaintenanceWork> => {
    const response = await apiClient.patch(`/maintenance-work/${workId}/complete`, data);
    return response.data;
  },
};

